### PR TITLE
Implement http::server (#115)

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

**Comment by eile** [comment]
@favreau @rdumusc @bilgili FYI

**Comment by eile** [comment]
Retest this please.

**Comment by eile** [comment]
Doc updated: http://hbpvis.github.io/zeq-0.5/index.html

File: tests/http/server.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
+
```
Comment (tribal-tec) [inline tests/http/server.cpp]: test needs to be excluded if httpxx is not found?!

File: tests/http/server.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        return jsonPut == json;
+    }
+
+    std::string _toJSON() const final
+    {
+        return jsonGet;
+    }
+};
+
+class Socket
```
Comment (tribal-tec) [inline tests/http/server.cpp]: name it HTTPClient?

File: zeq/http/server.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    explicit Server( Server& shared )
+        : Server( static_cast< Receiver& >( shared )) {}
+    ZEQ_API virtual ~Server();
+
+    /**
+     * Create a new Server when requested.
+     *
+     * The creation and parameters depend on the following command line
+     * parameters:
+     * * --http-server [host][:port]: Enable the server. Optional parameters
```
Comment (tribal-tec) [inline zeq/http/server.h]: what are optional parameters?

File: zeq/http/server.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+     * To process requests on the incoming port, call receive().
+     *
+     * @param uri The server address in the form "[tcp://][hostname][:port]"
+     * @param shared a shared receiver, see Receiver constructor.
+     * @throw std::runtime_error on malformed URI or connection issues.
+     */
+    ZEQ_API Server( const URI& uri, Receiver& shared );
+    ZEQ_API explicit Server( const URI& uri );
+    ZEQ_API explicit Server( Receiver& shared );
+    ZEQ_API explicit Server();
```
Comment (tribal-tec) [inline zeq/http/server.h]: all ctors should have their own documentation, at least the default ctor that it runs on localhost:4020 for instance.

**Comment by eile** [comment]
Retest this please.

File: tests/http/server.cpp (hunk 1 above)
Comment (eile) [inline tests/http/server.cpp]: done

File: tests/http/server.cpp (hunk 2 above)
Comment (eile) [inline tests/http/server.cpp]: Named it Client

File: zeq/http/server.h (hunk 3 above)
Comment (eile) [inline zeq/http/server.h]: [host] and [:port]. Documented it better.

File: zeq/http/server.h (hunk 4 above)
Comment (eile) [inline zeq/http/server.h]: Done.

**Comment by eile** [comment]
Retest this please.

**Comment by eile** [comment]
Retest this please.

**Comment by eile** [comment]
Retest this please.

**Comment by eile** [comment]
Good to merge?

File: doc/feature/httpServer.md [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
-* Accept PUT and POST requests and update subscribed Zerobuf in the same
+* Accept POST requests to update subscribed Zerobuf in the same
   way as a zeq::Subscriber
 * Accept GET requests and reply with the current state of registered
   Zerobuf in the same way as a Publisher::publish().
 * Allow custom code execution while handling a GET request, e.g., to
   render a new frame before serving a ImageJPEG request
 * Implement the same client REST API as the current RESTBridge
+* Do not accept PUT requests (as RestBridge does for updates): They are
+  for adding objects, which we might introduce later.
```
Comment (favreau) [inline doc/feature/httpServer.md]: According to most documentation, POST is for idempotent actions, and PUT is for non-idempotent ones. One can understand it as POST=create and PUT=create or update. If there is one to keep, PUT appears more appropriate.

Comment (eile) [inline doc/feature/httpServer.md]: All right, will change this.

File: zeq/http/server.h [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+#include <zeq/receiver.h> // base class
+
+namespace zeq
+{
+/** HTTP protocol support. */
+namespace http
+{
+/**
+ * Serves HTTP GET and POST requests for servus::Serializable objects.
```
Comment (tribal-tec) [inline zeq/http/server.h]: no POST!

File: zeq/http/server.h [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+ *
+ * Behaves semantically like a Publisher (for GET) and Subscriber (for POST),
```
Comment (tribal-tec) [inline zeq/http/server.h]: dito

**Comment by tribal-tec** [comment]
+1

**Pull request merged**